### ISSUE TITLE
mldonkey-next gui doesn't connect to the core when accessing through a reverse proxy

### ISSUE THREAD

**Issue by luke-70it** [carlonluca/docker-mldonkey#11]
Hi,
I've just found your version of mldonkey, and i'm very interested in a modern responsive web gui.
I've stopped my previous container (logicwar/mldonkey) and I deployed yours, using same ports and added new ones specific for your version.
From my LAN it works flawless both using old gui (port 4080) and "next" gui (port 4081).

To connect from Internet, I use Swag container as reverse proxy. It listen on port 443 for a specific FQDN and translate to internal ip on port 4081. In this scenario I can't connect to mldonley-next gui, but only to old gui.

4002 port is exposed, but something is missing in my configuration to make it works in a reverse proxy environment.

Can you help me to solve this problem?

Thank you!
Luca

**Comment by carlonluca** [comment]
Using a proxy for the new GUI should be no different from doing it for the old GUI. If the webapp is not downloaded, you'll probably want to check the logs of your reverse proxy first.
Please note that the webapp will also need to connect to port 4002 to communicate with the mldonkey core.

**Comment by luke-70it** [comment]
Is 4002 port accessed using http protocol?
If i add to reverse proxy configuration for the same server_name a section for port 4002 and try to connect to http://server_name:4002 i see the text "upgrade required"
Is this the expected behaviour?
Telnet to port 4002 worked even before this nginx configuration, but new gui still can't connect to core.
If i expose 4081 port an bypass reverse proxy it works, but i prefer https connection on 443 port.

**Comment by carlonluca** [comment]
4002 is used for the WebSocket server. WebSocket is somehow compatible with HTTP and may be used with a reverse proxy, however this is not something I tried.

So I guess you know can download the new GUI through the reverse proxy. If you see the "can't connect to core" error, it means you still need to make the WebSocket server available somehow.

**Comment by raulfg3** [comment]
> 002 is used for the WebSocket server. WebSocket is somehow compatible with HTTP and may be used with a reverse proxy, however this is not something I tried.
> 
> So I guess you know can download the new GUI through the reverse proxy. If you see the "can't connect to core" error, it means you still need to make the WebSocket server available somehow.

sorry but I do not know how make webSocket server available, I use option in NPM 
![image](https://github.com/user-attachments/assets/7b15238f-3a94-4bbd-be7e-29dcdbb71b1f)
 but Do not work as expected, perhaps you can test NPM, is a docker really easy to install: https://hub.docker.com/r/jc21/nginx-proxy-manager

on github:  https://github.com/NginxProxyManager/nginx-proxy-manager

As I say previosly , I can reverse proxy mldonkey, but not mldonkey-next , allways show "not connected to the core" that is the main problem

**Comment by carlonluca** [comment]
For mldonkey-next you need to make port 4002 (and optionally 4003) available. Is that available?

**Comment by raulfg3** [comment]
yes.

this is my docker compose.

`services:
  mldonkey:
    image: carlonluca/mldonkey:dev
    container_name: mldonkey
    networks:
      my-net:
    environment:
      - MLDONKEY_UID=$PUID
      - MLDONKEY_GID=$PGID
      - TZ=$TZ
      - MLDONKEY_ADMIN_PASSWORD=$Password
    volumes:
      - CHANGE_TO_COMPOSE_DATA_PATH/Dockers/mldonkey:/var/lib/mldonkey
      - CHANGE_TO_COMPOSE_DATA_PATH/Descargas/mldonkey:/var/lib/mldonkey/incoming # SKIP_BACKUP
      - $Temp/mldonkey_tmp:/var/lib/mldonkey/mldonkey_tmp # SKIP_BACKUP
    ports:
      - 4000:4000
      - 4001:4001
      - 4002:4002
      - 4003:4003
      - 4080:4080
      - 4081:4081
      - 20562:20562
      - 20566:20566/udp
    restart: unless-stopped
networks:
  my-net:
    external: true`


and as I say if use http://myIP:4081 works fine and i'm able to conect to core & login, but do not work if use NPM as reverse proxy ( reverse proxy works for old mldonkey on port 4080.

can do you tell me how to test port 4002 & 4003 are available thought a reverse proxy?

**Comment by carlonluca** [comment]
Maybe you need to forward the port? I'm sorry, this is specific to your setup.

You can connect with any websocket client and see if it connects properly. Logs are still pretty verbose in the container. I will probably reduce logging soon. For example:

`websocat ws://<ip>:4002/`

results in these logs (docker logs):

```
 [LOGGER]  24-11-30 14:50:07  info : Client connected: ... [object Object]
 [LOGGER]  24-11-30 14:50:07  verbose : Clients still connected: 1
 [LOGGER]  24-11-30 14:50:07  info : Connected to: 127.0.0.1:4001
 [LOGGER]  24-11-30 14:50:07  verbose : Sending data to client
 [LOGGER]  24-11-30 14:50:10  verbose : Clients still connected: 0
 [LOGGER]  24-11-30 14:50:10  info : Connection closed
```

**Comment by luke-70it** [comment]
Hello Luca,
can you please tell me if websocket use ws or wss and what is the exact path? is it ws://address:4002/ or is there a subfolder (i.e. ws://addres:4002/websocket/)?
I used a chrome plugin and I tried to open a connection to ws://address:4002/. The result was "OPENED"...

**Comment by raulfg3** [comment]
more info; I notice that CAN connect if NPM is configured to use http instead of HTTPS to connect AND use Firefox or Microsoft EDGE, if use Chrome ( my default browser), alway fails because chrome try to connect to HTTPS:// regardless that i write Http:// in the side bar.

So conclusion: works for http:// and core connect on browser that are NOT Chrome

**Comment by raulfg3** [comment]
Yes, I want to use it from my work, mldonkey run on my home NAS and I have NGINX Proxy Manager to share several services to internet so I Can Acess it from outside my home eg: I use Kavita, Airsonic , transmision, and of course mldonkey, all of then use paswword to acces it, and https:// to access it , except mldonkey next core that only works on https:// and on microsoft edge.

PD: Old mldonkey webGUI (4080) is accesible by https:// so I actually use it, but I want to use webGUI from donkey next (4081) if possible.

**Comment by raulfg3** [comment]
I have interest, if is not too work, please consider to implement it.

Thanks in advanced.

**Comment by carlonluca** [comment]
I added some changes to make this setup simpler in the new branch https://github.com/carlonluca/docker-mldonkey/tree/reverse-proxy. I also updated the readme with examples.

I tested this new branch quickly and it seemed to work: I can access the mldonkey-next webapp throught nginx with a custom public domain name and a proper SSL certificate. As said, I do not consider this setup safe though.

Note that this branch simplifies the port setup: 4002 and 4003 are no more needed. Websocket servers are now listening on the same port of the webapp (4081). If you want to log stream, you'll also need to set an env variable (it is unsafe).

This version of the image is the one with tag rev-proxy in docker hub. It should work with HTTPS, with HTTP and without reverse proxy.

NOTE1: this image uses the new branch with ocaml5, which is still experimental. Please report bugs if you find any.
NOTE2: I found (and hopefully fixed) a bug in the native app for which a hostname could not be used. I'll have to release an update if you want to use the Android app from the Internet. The current version won't probably work.

Please let me know what you think. I'll probably merge to master in the next few days.

**Comment by carlonluca** [comment]
The latest tag will soon be updated with this new version including ocaml5/https/rev proxy support.
The native app was updated and should now support the new setup.

**Comment by raulfg3** [comment]
good, news, I'll wait some days and change tag to latest on my docker, to test it, I write here test result.


Thanks a lot
